Part2-Euclid/main.cpp

```diff
@@ -21,12 +44,288 @@ int main(int argc, char** argv) {
 
-	std::array<sample, CLASSES> samples;
+	sample misclassifications[CLASSES];
+	array<sample, CLASSES> samples;
+	array<double, CLASSES> logPriors, varDets = {}, logVarDets = {};
+	array<CovMatrix, CLASSES> varInverses;
+	array<observation, CLASSES> means = getMeans(arg.set);
+	array<CovMatrix, CLASSES> vars    = getVars(arg.set);
+	array<unsigned, CLASSES> sizes    = getSizes(arg.set);
 	getSamples(arg.set, samples, arg.seed);
 
+	observation min = samples[0].front(), max = samples[0].front();
+
+	double totalSize = std::accumulate(sizes.begin(), sizes.end(), 0);
+
+	// Compute the log of the priors
+	std::transform(sizes.cbegin(), sizes.cend(), logPriors.begin(),
+	               [totalSize](unsigned size) { return log(size / totalSize); });
+
+	// Compute which case we're in from the book
+	if (arg.discriminantCase == 0) {
+		arg.discriminantCase = detectCase(vars);
+
+		std::cout << "Detected case " << arg.discriminantCase << "\n\n";
+	} else {
+		std::cout << "Overriden case " << arg.discriminantCase << "\n\n";
+	}
+
+	calcInversesAndDets(arg.discriminantCase, vars, varInverses, varDets, !arg.pdfPlotFile.fail());
+
+	// Calculate the logs of the determinants
+	std::transform(varDets.cbegin(), varDets.cend(), logVarDets.begin(), [](const double& det) { return log(det); });
+
+	unsigned overallMisclass = 0;
+
+	for (unsigned i = 0; i < CLASSES; i++) {
+		sample& misclass = misclassifications[i];
+
+		unsigned misclassCount =
+		    classifySample(arg.discriminantCase, samples[i], i, misclassifications[i], min, max, means, varInverses,
+		                   logVarDets, logPriors, !arg.misclassPlotFiles[i].fail());
+
+		std::cout << "Misclassification rate for class " << i + 1 << ":\n"
+		          << misclassCount / (double) sizes[i] << "\n\n";
+
+		overallMisclass += misclassCount;
+
+		if (arg.plotFiles[i]) { printPlotFile(arg.plotFiles[i], samples[i]); }
+
+		if (arg.misclassPlotFiles[i]) { printPlotFile(arg.misclassPlotFiles[i], misclassifications[i]); }
+	}
+
+	std::cout << "Overall misclassification rate:\n" << overallMisclass / totalSize << '\n';
+
+	double pdfMax = 0;
+	if (arg.pdfPlotFile) {
+		pdfMax = printPdfPlotFile(arg.pdfPlotFile, min, max, means, varInverses, varDets, logPriors);
+
+		// Round it so that the z tick values are not so ugly
+		pdfMax = round(pdfMax * 100) / 100;
+	}
+
+	if (arg.boundaryParamsFile) {
+		printParamsFile(arg.boundaryParamsFile, min, max, means, varInverses, logVarDets, logPriors, pdfMax);
+	}
+
 	return 0;
 }
 
+double discriminate(const observation& obs, const Vec<CLASSES>& mu, const CovMatrix& varInverse, double logVarDet,
+                         double logPrior) {
+	// Note that varInverse is what is passed, and the inverse of a diagonal matrix is also a
+	// diagonal matrix with the diagonal elements being the reciprocal of the original elements. So
+	// instead of dividing by sigma^2, we multiply by 1/sigma^2.
+	return (varInverse(0, 0) * mu).dot(obs) - varInverse(0, 0) * mu.dot(mu) / 2.0;
+}
```

Comment:
Consider changing function signature to remove logVarDet and logPrior (unneeded) and pass in sigma^2 instead of varInverse, since function signature doesn't have to match other cases.